Sources/PackageLoading/TargetSourcesBuilder.swift

```diff
@@ -0,0 +1,311 @@
+                return Resource(rule: .process, path: $0.key)
+            case .copy:
+                return Resource(rule: .copy, path: $0.key)
+            }
+        }
+
+        return (sources, resources)
+    }
+
+    /// Find the rule for the given path.
+    private func findRule(for path: AbsolutePath) -> FileRuleDescription.Rule {
+        var matchedRule: FileRuleDescription.Rule = .none
+
+        // First match any resources explicitly declared in the manifest file.
+        for declaredResource in target.resources {
+            let resourcePath = self.targetPath.appending(RelativePath(declaredResource.path))
+            if path.contains(resourcePath) {
+                if matchedRule != .none {
+                    diags.emit(.error("Duplicate rule \(declaredResource.rule) found for \(path)"))
+                }
+                matchedRule = declaredResource.rule.fileRule
+            }
+        }
+
+        // Match any sources explicitly declared in the manifest file.
+        if let declaredSources = target.sources {
+            for declaredSource in declaredSources {
+                let sourcePath = self.targetPath.appending(RelativePath(declaredSource))
+                if path.contains(sourcePath) {
+                    if matchedRule != .none {
+                        diags.emit(.error("Duplicate rule compile found for \(path)"))
+                    }
+                    matchedRule = .compile
+                }
+            }
+        }
+
+        // We haven't found a rule using that's explicitly declared in the manifest
+        // so try doing an automatic match.
+        if matchedRule == .none {
+            let effectiveRules: [FileRuleDescription] = {
+                // Don't automatically match compile rules if target's sources are
+                // explicitly declared in the package manifest.
+                if target.sources != nil {
+                    return self.rules.filter { $0.rule == .compile }
+                }
+                return self.rules
+            }()
+
+            if let needle = effectiveRules.first(where: { $0.match(path: path) }) {
+                matchedRule = needle.rule
+            }
+        }
+
+        return matchedRule
+    }
+
+    /// Compute the contents of the files in a target.
+    ///
+    /// This avoids recursing into certain directories like exclude or the
+    /// ones that should be copied as-is.
+    public func computeContents() -> [AbsolutePath] {
+        var contents: [AbsolutePath] = []
+        var queue: [AbsolutePath] = [targetPath]
+
+        while let curr = queue.popLast() {
+            // Ignore dot files.
+            if curr.basename.hasPrefix(".") { continue }
+
+            // Ignore xcodeproj and playground directories.
+            //
+            // FIXME: Ignore lproj directories until we add localization support.
+            if ["xcodeproj", "playgroun", "lproj"].contains(curr.extension) { continue }
```

Comment:
missing "d" in playground, should this also do xcworkspace?

Reply:
yeah, it should.